libraries/Microsoft.Bot.Builder.Ai/TranslationMiddleware.cs

```diff
@@ -40,48 +46,33 @@ public TranslationMiddleware(HttpClient httpClient, string[] nativeLanguages, st
                 if (!String.IsNullOrWhiteSpace(message.Text))
                 {
                     // determine the language we are using for this conversation
-                    var sourceLanguage = "en"; // context.Conversation.Data["Language"]?.ToString() ?? this.nativeLanguages.FirstOrDefault() ?? "en";
-
-                    var translationContext = new TranslationContext
+                    var sourceLanguage = await Task.Run(() => translator.Detect(message.Text)); // context.Conversation.Data["Language"]?.ToString() ?? this.nativeLanguages.FirstOrDefault() ?? "en";
+                    var templatePath = Path.Combine(new string[] { templatesDir, sourceLanguage + ".template" });
+                    if (File.Exists(templatePath))
+                    {
+                        this.translator.SetPostProcessorTemplate(templatePath);
+                    }
+                    if (!nativeLanguages.Contains(sourceLanguage))
```

Comment:
>nativeLanguages.Contains(sourceLanguage) [](start = 25, length = 40)

what happens if someone adds your middleware twice? Seems like you should check for that by looking for your "Translation" context and reporting an error if you detect you've been added twice.